File: apps/apple/core/Sources/LorvexWorkflow/TaskChecklist.swift
import Foundation
import GRDB
import LorvexDomain
import LorvexStore

/// Per-item checklist mutations: add, update text, toggle completion,
/// remove, reorder. Each op stamps a fresh `version` + `updated_at`
/// on every touched child row so peers' LWW reconciliation accepts the change.
public enum TaskChecklist {
  /// Sync-fanout operation applied to a single checklist item.
  public enum ChecklistSyncOperation: Sendable, Equatable {
    case upsert
    case delete

    public var asString: String {
      switch self {
      case .upsert: return SyncNaming.opUpsert
      case .delete: return SyncNaming.opDelete
      }
    }
  }

  /// One sync-fanout change record produced by a checklist mutation.
  /// `snapshot` is populated only for delete operations, carrying the
  /// pre-delete payload the sync layer enqueues as a tombstone body.
  public struct ChecklistItemSyncChange: Sendable, Equatable {
    public let itemId: String
    public let operation: ChecklistSyncOperation
    public let snapshot: JSONValue?
    public let version: String
  }

  /// Rich return shape from every checklist mutation: pre/post task
  /// JSON, a human-readable summary, and the per-item sync changes.
  public struct MutationResult: Sendable {
    public let taskId: String
    public let beforeTask: JSONValue
    public let afterTask: JSONValue
    public let summary: String
    public let itemSyncChanges: [ChecklistItemSyncChange]
  }

  // MARK: - Inputs

  public struct AddInput: Sendable {
    public let taskId: TaskId
    public let text: String
    public let position: UInt32?
    public init(taskId: TaskId, text: String, position: UInt32? = nil) {
      self.taskId = taskId
      self.text = text
      self.position = position
    }
  }

  public struct UpdateInput: Sendable {
    public let itemId: ChecklistItemId
    public let text: String
    public init(itemId: ChecklistItemId, text: String) {
      self.itemId = itemId
      self.text = text
    }
  }

  public struct ToggleInput: Sendable {
    public let itemId: ChecklistItemId
    public let completed: Bool
    public init(itemId: ChecklistItemId, completed: Bool) {
      self.itemId = itemId
      self.completed = completed
    }
  }

  public struct RemoveInput: Sendable {
    public let itemId: ChecklistItemId
    public init(itemId: ChecklistItemId) { self.itemId = itemId }
  }

  public struct ReorderInput: Sendable {
    public let taskId: TaskId
    public let itemIds: [ChecklistItemId]
    public init(taskId: TaskId, itemIds: [ChecklistItemId]) {
      self.taskId = taskId
      self.itemIds = itemIds
    }
  }

  // MARK: - Operations

  /// Insert a checklist item at `input.position` (or the end when
  /// omitted), re-stamping every sibling row's `position`.
  public static func addTaskChecklistItem(
    _ db: Database, hlc: HlcSession, input: AddInput
  ) throws -> MutationResult {
    let text = UnicodeHygiene.sanitizeUserText(input.text)
    do { try validateTaskChecklistItemText(text) } catch let e as ValidationError {
      throw StoreError.validation(e.description)
    }

    let before = try TaskResponse.loadEnrichedTaskJSON(db, taskId: input.taskId)
    let title = TaskResponse.taskTitle(before)
    let existingItems = try fetchChecklistItemsForTask(db, taskId: input.taskId)
    do { try validateTaskChecklistItemCount(existingItems.count + 1) }
    catch let e as ValidationError {
      throw StoreError.validation(e.description)
    }

    let insertIndex: Int
    if let p = input.position {
      insertIndex = Int(p)
    } else {
      insertIndex = existingItems.count
    }
    if insertIndex > existingItems.count {
      throw StoreError.validation(
        "checklist insert position \(insertIndex) is out of range for task '\(input.taskId.asString)' with \(existingItems.count) items"
      )
    }

    let now = SyncTimestampFormat.syncTimestampNow()
    let newItemId = ChecklistItemId.new()
    var ordered = existingItems.map(\.itemId)
    ordered.insert(newItemId, at: insertIndex)
    let versionsById = Dictionary(
      uniqueKeysWithValues: existingItems.map { ($0.itemId.rawValue, $0.version) })

    var changes: [ChecklistItemSyncChange] = []
    changes.reserveCapacity(ordered.count)
    for (index, itemId) in ordered.enumerated() {
      let version = try VersionFloor.mint(
        hlc: hlc,
        existingVersion: versionsById[itemId.rawValue],
        entityType: EntityName.taskChecklistItem,
        entityId: itemId.rawValue)
      if itemId.rawValue == newItemId.rawValue {
        try db.execute(
          sql:
            "INSERT INTO task_checklist_items "
            + "(id, task_id, position, text, completed_at, version, created_at, updated_at) "
            + "VALUES (?, ?, ?, ?, NULL, ?, ?, ?)",
          arguments: [
            newItemId.rawValue, input.taskId.rawValue, Int64(index), text,
            version, now, now,
          ])
      } else {
        let existingVersion = versionsById[itemId.rawValue]
        try db.execute(
          sql:
            "UPDATE task_checklist_items "
            + "SET position = ?, version = ?, updated_at = ? "
            + "WHERE id = ? AND version = ?",
          arguments: [Int64(index), version, now, itemId.rawValue, existingVersion])
        try requireChecklistItemUpdate(
          db,
          itemId: itemId,
          attemptedVersion: version)
      }
      changes.append(itemUpsertChange(itemId, version: version))
    }

    let after = try TaskResponse.loadEnrichedTaskJSON(db, taskId: input.taskId)
    return MutationResult(
      taskId: input.taskId.asString,
      beforeTask: before,
      afterTask: after,
      summary: "Added checklist item '\(text)' for '\(title)'",
      itemSyncChanges: changes)
  }

  /// Rewrite an item's text. Sanitized + validated.
  public static func updateTaskChecklistItem(
    _ db: Database, hlc: HlcSession, input: UpdateInput
  ) throws -> MutationResult {
    let text = UnicodeHygiene.sanitizeUserText(input.text)
    do { try validateTaskChecklistItemText(text) } catch let e as ValidationError {
      throw StoreError.validation(e.description)
    }

    let identity = try fetchChecklistItemIdentity(db, itemId: input.itemId)
    let before = try TaskResponse.loadEnrichedTaskJSON(db, taskId: identity.taskId)
    let title = TaskResponse.taskTitle(before)
    let now = SyncTimestampFormat.syncTimestampNow()
    let version = try VersionFloor.mint(
      hlc: hlc,
      existingVersion: identity.version,
      entityType: EntityName.taskChecklistItem,
      entityId: input.itemId.rawValue)
    try db.execute(
      sql:
        "UPDATE task_checklist_items "
        + "SET text = ?, version = ?, updated_at = ? "
        + "WHERE id = ? AND version = ?",
      arguments: [text, version, now, input.itemId.rawValue, identity.version])
    try requireChecklistItemUpdate(db, itemId: input.itemId, attemptedVersion: version)

    let after = try TaskResponse.loadEnrichedTaskJSON(db, taskId: identity.taskId)
    return MutationResult(
      taskId: identity.taskId.asString,
      beforeTask: before,
      afterTask: after,
      summary:
        "Updated checklist item '\(identity.text)' for '\(title)'",
      itemSyncChanges: [itemUpsertChange(input.itemId, version: version)])
  }

  /// Toggle `completed_at` (sets it to `now` when completing, NULL when
  /// reopening).
  public static func toggleTaskChecklistItem(
    _ db: Database, hlc: HlcSession, input: ToggleInput
  ) throws -> MutationResult {
    let identity = try fetchChecklistItemIdentity(db, itemId: input.itemId)
    let before = try TaskResponse.loadEnrichedTaskJSON(db, taskId: identity.taskId)
    let title = TaskResponse.taskTitle(before)
    let now = SyncTimestampFormat.syncTimestampNow()
    let version = try VersionFloor.mint(
      hlc: hlc,
      existingVersion: identity.version,
      entityType: EntityName.taskChecklistItem,
      entityId: input.itemId.rawValue)
    let completedAt: String? = input.completed ? now : nil
    try db.execute(
      sql:
        "UPDATE task_checklist_items "
        + "SET completed_at = ?, version = ?, updated_at = ? "
        + "WHERE id = ? AND version = ?",
      arguments: [completedAt, version, now, input.itemId.rawValue, identity.version])
    try requireChecklistItemUpdate(db, itemId: input.itemId, attemptedVersion: version)

    let action = input.completed ? "Completed" : "Reopened"
    let after = try TaskResponse.loadEnrichedTaskJSON(db, taskId: identity.taskId)
    return MutationResult(
      taskId: identity.taskId.asString,
      beforeTask: before,
      afterTask: after,
      summary: "\(action) checklist item '\(identity.text)' for '\(title)'",
      itemSyncChanges: [itemUpsertChange(input.itemId, version: version)])
  }

  /// Remove an item; restamps every remaining sibling's `position`.
  public static func removeTaskChecklistItem(
    _ db: Database, hlc: HlcSession, input: RemoveInput
  ) throws -> MutationResult {
    let identity = try fetchChecklistItemIdentity(db, itemId: input.itemId)
    let preDeleteSnapshot = try PayloadLoaders.loadTaskChecklistItemSyncPayload(
      db, itemId: input.itemId.rawValue)
    let before = try TaskResponse.loadEnrichedTaskJSON(db, taskId: identity.taskId)
    let title = TaskResponse.taskTitle(before)
    let now = SyncTimestampFormat.syncTimestampNow()
    let deleteVersion = try VersionFloor.mint(
      hlc: hlc,
      existingVersion: identity.version,
      entityType: EntityName.taskChecklistItem,
      entityId: input.itemId.rawValue)
    let deletedSnapshot = preDeleteSnapshot.map {
      payload($0, replacingVersionWith: deleteVersion)
    }

    try db.execute(
      sql: "DELETE FROM task_checklist_items WHERE id = ? AND version = ?",
      arguments: [input.itemId.rawValue, identity.version])
    try requireChecklistItemDelete(
      db,
      itemId: input.itemId,
      attemptedVersion: deleteVersion)

    let remaining = try fetchChecklistItemsForTask(db, taskId: identity.taskId)
    var siblingChanges: [ChecklistItemSyncChange] = []
    siblingChanges.reserveCapacity(remaining.count)
    for (index, item) in remaining.enumerated() {
      let version = try VersionFloor.mint(
        hlc: hlc,
        existingVersion: item.version,
        entityType: EntityName.taskChecklistItem,
        entityId: item.itemId.rawValue)
      try db.execute(
        sql:
          "UPDATE task_checklist_items "
          + "SET position = ?, version = ?, updated_at = ? "
          + "WHERE id = ? AND version = ?",
        arguments: [Int64(index), version, now, item.itemId.rawValue, item.version])
      try requireChecklistItemUpdate(
        db,
        itemId: item.itemId,
        attemptedVersion: version)
      siblingChanges.append(itemUpsertChange(item.itemId, version: version))
    }

    let after = try TaskResponse.loadEnrichedTaskJSON(db, taskId: identity.taskId)
    var changes: [ChecklistItemSyncChange] = []
    changes.reserveCapacity(remaining.count + 1)
    changes.append(
      ChecklistItemSyncChange(
        itemId: input.itemId.rawValue,
        operation: .delete,
        snapshot: deletedSnapshot,
        version: deleteVersion))
    changes.append(contentsOf: siblingChanges)

    return MutationResult(
      taskId: identity.taskId.asString,
      beforeTask: before,
      afterTask: after,
      summary: "Removed checklist item '\(identity.text)' for '\(title)'",
      itemSyncChanges: changes)
  }

  /// Reorder every item in one call. `itemIds` must contain every
  /// existing checklist item for `input.taskId` exactly once.
  public static func reorderTaskChecklistItems(
    _ db: Database, hlc: HlcSession, input: ReorderInput
  ) throws -> MutationResult {
    let before = try TaskResponse.loadEnrichedTaskJSON(db, taskId: input.taskId)
    let title = TaskResponse.taskTitle(before)
    let existing = try fetchChecklistItemsForTask(db, taskId: input.taskId)

    if input.itemIds.count != existing.count {
      throw StoreError.validation(
        "reorder_task_checklist_items requires exactly \(existing.count) ids for task '\(input.taskId.asString)', got \(input.itemIds.count)"
      )
    }
    let existingSet = Set(existing.map(\.itemId.rawValue))
    let requestedSet = Set(input.itemIds.map(\.rawValue))
    if existingSet != requestedSet || requestedSet.count != input.itemIds.count {
      throw StoreError.validation(
        "reorder_task_checklist_items must contain every checklist item for task '\(input.taskId.asString)' exactly once"
      )
    }

    let now = SyncTimestampFormat.syncTimestampNow()
    let versionsById = Dictionary(
      uniqueKeysWithValues: existing.map { ($0.itemId.rawValue, $0.version) })
    var changes: [ChecklistItemSyncChange] = []
    changes.reserveCapacity(input.itemIds.count)
    for (index, itemId) in input.itemIds.enumerated() {
      let existingVersion = versionsById[itemId.rawValue]
      let version = try VersionFloor.mint(
        hlc: hlc,
        existingVersion: existingVersion,
        entityType: EntityName.taskChecklistItem,
        entityId: itemId.rawValue)
      try db.execute(
        sql:
          "UPDATE task_checklist_items "
          + "SET position = ?, version = ?, updated_at = ? "
          + "WHERE id = ? AND task_id = ? AND version = ?",
        arguments: [
          Int64(index), version, now, itemId.rawValue, input.taskId.rawValue,
          existingVersion,
        ])
      try requireChecklistItemUpdate(db, itemId: itemId, attemptedVersion: version)
      changes.append(itemUpsertChange(itemId, version: version))
    }

    let after = try TaskResponse.loadEnrichedTaskJSON(db, taskId: input.taskId)
    return MutationResult(
      taskId: input.taskId.asString,
      beforeTask: before,
      afterTask: after,
      summary: "Reordered checklist items for '\(title)'",
      itemSyncChanges: changes)
  }

  // MARK: - Helpers

  /// Lightweight identity row read from a single SELECT, used by every
  /// per-item op to look up the owning task + display text for the
  /// summary string.
  struct ItemIdentity {
    let taskId: TaskId
    let text: String
    let position: Int64
    let completed: Bool
    let version: String
  }

  static func fetchChecklistItemIdentity(
    _ db: Database, itemId: ChecklistItemId
  ) throws -> ItemIdentity {
    guard
      let row = try Row.fetchOne(
        db,
        sql:
          "SELECT task_id, text, position, completed_at IS NOT NULL, version "
          + "FROM task_checklist_items WHERE id = ?",
        arguments: [itemId.rawValue])
    else {
      throw StoreError.notFound(
        entity: "checklist item", id: itemId.rawValue)
    }
    return ItemIdentity(
      taskId: TaskId(trusted: row[0]),
      text: row[1],
      position: row[2],
      completed: (row[3] as Int64) != 0,
      version: row[4])
  }

  struct VersionedItem {
    let itemId: ChecklistItemId
    let version: String
  }

  static func fetchChecklistItemsForTask(
    _ db: Database, taskId: TaskId
  ) throws -> [VersionedItem] {
    let rows = try Row.fetchAll(
      db,
      sql:
        "SELECT id, version FROM task_checklist_items "
        + "WHERE task_id = ? "
        + "ORDER BY position ASC, created_at ASC, id ASC",
      arguments: [taskId.rawValue])
    return rows.map {
      VersionedItem(itemId: ChecklistItemId(trusted: $0[0]), version: $0[1])
    }
  }

  static func requireChecklistItemUpdate(
    _ db: Database, itemId: ChecklistItemId, attemptedVersion: String
  ) throws {
    guard db.changesCount == 0 else { return }
    guard
      let winner = try String.fetchOne(
        db, sql: "SELECT version FROM task_checklist_items WHERE id = ?",
        arguments: [itemId.rawValue])
    else {
      throw StoreError.notFound(entity: EntityName.taskChecklistItem, id: itemId.rawValue)
    }
    throw StoreError.versionSuperseded(
      entityType: EntityName.taskChecklistItem,
      entityId: itemId.rawValue,
      attemptedVersion: attemptedVersion,
      existingVersion: winner)
  }

  static func requireChecklistItemDelete(
    _ db: Database, itemId: ChecklistItemId, attemptedVersion: String
  ) throws {
    guard db.changesCount == 0 else { return }
    guard
      let winner = try String.fetchOne(
        db, sql: "SELECT version FROM task_checklist_items WHERE id = ?",
        arguments: [itemId.rawValue])
    else {
      throw StoreError.notFound(entity: EntityName.taskChecklistItem, id: itemId.rawValue)
    }
    throw StoreError.versionSuperseded(
      entityType: EntityName.taskChecklistItem,
      entityId: itemId.rawValue,
      attemptedVersion: attemptedVersion,
      existingVersion: winner)
  }

  static func payload(_ payload: JSONValue, replacingVersionWith version: String) -> JSONValue {
    guard case .object(var object) = payload else { return payload }
    object["version"] = .string(version)
    return .object(object)
  }

  static func itemUpsertChange(
    _ itemId: ChecklistItemId, version: String
  ) -> ChecklistItemSyncChange {
    ChecklistItemSyncChange(
      itemId: itemId.rawValue, operation: .upsert, snapshot: nil, version: version)
  }
}

File: apps/tauri/lorvex-workflow/src/task_checklist.rs
use std::collections::HashSet;

use lorvex_domain::hlc_session::HlcSession;
use lorvex_domain::naming::{ENTITY_TASK, OP_DELETE, OP_UPSERT};
use lorvex_domain::{
    validate_task_checklist_item_count, validate_task_checklist_item_text, ChecklistItemId, TaskId,
};
use lorvex_store::payload_loaders::load_task_checklist_item_sync_payload;
use lorvex_store::StoreError;
use rusqlite::{params, Connection};
use serde_json::Value;

use crate::task_response::{load_enriched_task_json, task_title};

#[derive(Debug, Clone)]
pub struct AddTaskChecklistItemInput {
    pub task_id: TaskId,
    pub text: String,
    pub position: Option<u32>,
}

#[derive(Debug, Clone)]
pub struct UpdateTaskChecklistItemInput {
    pub item_id: ChecklistItemId,
    pub text: String,
}

#[derive(Debug, Clone)]
pub struct ToggleTaskChecklistItemInput {
    pub item_id: ChecklistItemId,
    pub completed: bool,
}

#[derive(Debug, Clone)]
pub struct RemoveTaskChecklistItemInput {
    pub item_id: ChecklistItemId,
}

#[derive(Debug, Clone)]
pub struct ReorderTaskChecklistItemsInput {
    pub task_id: TaskId,
    pub item_ids: Vec<ChecklistItemId>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ChecklistSyncOperation {
    Upsert,
    Delete,
}

impl ChecklistSyncOperation {
    #[must_use]
    pub const fn as_str(&self) -> &'static str {
        match self {
            Self::Upsert => OP_UPSERT,
            Self::Delete => OP_DELETE,
        }
    }
}

#[derive(Debug, Clone)]
pub struct ChecklistItemSyncChange {
    pub item_id: String,
    pub operation: ChecklistSyncOperation,
    pub snapshot: Option<Value>,
}

#[derive(Debug, Clone)]
pub struct TaskChecklistMutationResult {
    pub task_id: String,
    pub before_task: Value,
    pub after_task: Value,
    pub summary: String,
    pub item_sync_changes: Vec<ChecklistItemSyncChange>,
}

fn fetch_checklist_item_identity(
    conn: &Connection,
    item_id: &ChecklistItemId,
) -> Result<(TaskId, String, i64, bool), StoreError> {
    conn.prepare_cached(
        "SELECT task_id, text, position, completed_at IS NOT NULL
         FROM task_checklist_items WHERE id = ?1",
    )?
    .query_row(params![item_id], |row| {
        let task_id: String = row.get(0)?;
        Ok((
            TaskId::from_trusted(task_id),
            row.get(1)?,
            row.get(2)?,
            row.get(3)?,
        ))
    })
    .map_err(|error| match error {
        rusqlite::Error::QueryReturnedNoRows => StoreError::NotFound {
            entity: "checklist item",
            id: item_id.to_string(),
        },
        other => StoreError::from(other),
    })
}

fn fetch_checklist_ids_for_task(
    conn: &Connection,
    task_id: &TaskId,
) -> Result<Vec<ChecklistItemId>, StoreError> {
    let mut stmt = conn.prepare_cached(
        "SELECT id
         FROM task_checklist_items
         WHERE task_id = ?1
         ORDER BY position ASC, created_at ASC, id ASC",
    )?;
    let ids = stmt
        .query_map(params![task_id], |row| {
            let id: String = row.get(0)?;
            Ok(ChecklistItemId::from_trusted(id))
        })?
        .collect::<Result<Vec<_>, _>>()?;
    Ok(ids)
}

fn touch_task_lww(
    conn: &Connection,
    task_id: &TaskId,
    version: &str,
    now: &str,
) -> Result<(), StoreError> {
    match conn
        .prepare_cached(
            "UPDATE tasks
         SET version = ?1, updated_at = ?2
         WHERE id = ?3 AND ?1 > version
         RETURNING 1",
        )?
        .query_row(params![version, now, task_id], |_row| Ok(()))
    {
        Ok(()) => Ok(()),
        Err(rusqlite::Error::QueryReturnedNoRows) => Err(StoreError::StaleVersion {
            entity: ENTITY_TASK,
            id: task_id.to_string(),
        }),
        Err(error) => Err(StoreError::from(error)),
    }
}

fn item_upsert_change(item_id: &ChecklistItemId) -> ChecklistItemSyncChange {
    ChecklistItemSyncChange {
        item_id: item_id.to_string(),
        operation: ChecklistSyncOperation::Upsert,
        snapshot: None,
    }
}

pub fn add_task_checklist_item(
    conn: &Connection,
    hlc: &HlcSession<'_>,
    input: AddTaskChecklistItemInput,
) -> Result<TaskChecklistMutationResult, StoreError> {
    let AddTaskChecklistItemInput {
        task_id,
        text,
        position,
    } = input;
    let text = lorvex_domain::sanitize_user_text(&text);
    validate_task_checklist_item_text(&text)?;

    let before = load_enriched_task_json(conn, &task_id)?;
    let title = task_title(&before).to_string();
    let existing_ids = fetch_checklist_ids_for_task(conn, &task_id)?;
    validate_task_checklist_item_count(existing_ids.len() + 1)?;

    let insert_index = position.map_or(existing_ids.len(), |value| value as usize);
    if insert_index > existing_ids.len() {
        return Err(StoreError::Validation(format!(
            "checklist insert position {insert_index} is out of range for task '{task_id}' with {} items",
            existing_ids.len()
        )));
    }

    let now = lorvex_domain::sync_timestamp_now();
    let item_id = ChecklistItemId::new();
    let mut ordered_ids = existing_ids;
    ordered_ids.insert(insert_index, item_id.clone());

    let parent_version = hlc.next_version_string();
    touch_task_lww(conn, &task_id, &parent_version, &now)?;

    let mut update_stmt = conn.prepare_cached(
        "UPDATE task_checklist_items
         SET position = ?1, version = ?2, updated_at = ?3
         WHERE id = ?4",
    )?;
    let mut insert_stmt = conn.prepare_cached(
        "INSERT INTO task_checklist_items (
            id, task_id, position, text, completed_at, version, created_at, updated_at
         ) VALUES (?1, ?2, ?3, ?4, NULL, ?5, ?6, ?6)",
    )?;

    for (index, existing_item_id) in ordered_ids.iter().enumerate() {
        let version = hlc.next_version_string();
        if existing_item_id == &item_id {
            insert_stmt.execute(params![item_id, task_id, index as i64, text, version, now])?;
        } else {
            update_stmt.execute(params![index as i64, version, now, existing_item_id])?;
        }
    }

    let after = load_enriched_task_json(conn, &task_id)?;
    let changed = ordered_ids.iter().map(item_upsert_change).collect();

    Ok(TaskChecklistMutationResult {
        task_id: task_id.to_string(),
        before_task: before,
        after_task: after,
        summary: format!("Added checklist item '{text}' for '{title}'"),
        item_sync_changes: changed,
    })
}

pub fn update_task_checklist_item(
    conn: &Connection,
    hlc: &HlcSession<'_>,
    input: UpdateTaskChecklistItemInput,
) -> Result<TaskChecklistMutationResult, StoreError> {
    let UpdateTaskChecklistItemInput { item_id, text } = input;
    let text = lorvex_domain::sanitize_user_text(&text);
    validate_task_checklist_item_text(&text)?;

    let (task_id, previous_text, _position, _completed) =
        fetch_checklist_item_identity(conn, &item_id)?;
    let before = load_enriched_task_json(conn, &task_id)?;
    let title = task_title(&before).to_string();
    let now = lorvex_domain::sync_timestamp_now();
    let parent_version = hlc.next_version_string();
    touch_task_lww(conn, &task_id, &parent_version, &now)?;
    let version = hlc.next_version_string();

    conn.prepare_cached(
        "UPDATE task_checklist_items
         SET text = ?1, version = ?2, updated_at = ?3
         WHERE id = ?4",
    )?
    .execute(params![text, version, now, item_id])?;

    let after = load_enriched_task_json(conn, &task_id)?;
    Ok(TaskChecklistMutationResult {
        task_id: task_id.to_string(),
        before_task: before,
        after_task: after,
        summary: format!("Updated checklist item '{previous_text}' for '{title}'"),
        item_sync_changes: vec![item_upsert_change(&item_id)],
    })
}

pub fn toggle_task_checklist_item(
    conn: &Connection,
    hlc: &HlcSession<'_>,
    input: ToggleTaskChecklistItemInput,
) -> Result<TaskChecklistMutationResult, StoreError> {
    let ToggleTaskChecklistItemInput { item_id, completed } = input;
    let (task_id, text, _position, _was_completed) = fetch_checklist_item_identity(conn, &item_id)?;
    let before = load_enriched_task_json(conn, &task_id)?;
    let title = task_title(&before).to_string();
    let now = lorvex_domain::sync_timestamp_now();
    let parent_version = hlc.next_version_string();
    touch_task_lww(conn, &task_id, &parent_version, &now)?;
    let version = hlc.next_version_string();
    let completed_at = completed.then(|| now.clone());

    conn.prepare_cached(
        "UPDATE task_checklist_items
         SET completed_at = ?1, version = ?2, updated_at = ?3
         WHERE id = ?4",
    )?
    .execute(params![completed_at, version, now, item_id])?;

    let action = if completed { "Completed" } else { "Reopened" };
    let after = load_enriched_task_json(conn, &task_id)?;
    Ok(TaskChecklistMutationResult {
        task_id: task_id.to_string(),
        before_task: before,
        after_task: after,
        summary: format!("{action} checklist item '{text}' for '{title}'"),
        item_sync_changes: vec![item_upsert_change(&item_id)],
    })
}

pub fn remove_task_checklist_item(
    conn: &Connection,
    hlc: &HlcSession<'_>,
    input: RemoveTaskChecklistItemInput,
) -> Result<TaskChecklistMutationResult, StoreError> {
    let RemoveTaskChecklistItemInput { item_id } = input;
    let (task_id, text, _position, _completed) = fetch_checklist_item_identity(conn, &item_id)?;
    let deleted_snapshot = load_task_checklist_item_sync_payload(conn, &item_id)?;
    let before = load_enriched_task_json(conn, &task_id)?;
    let title = task_title(&before).to_string();
    let now = lorvex_domain::sync_timestamp_now();
    let parent_version = hlc.next_version_string();
    touch_task_lww(conn, &task_id, &parent_version, &now)?;

    conn.prepare_cached("DELETE FROM task_checklist_items WHERE id = ?1")?
        .execute(params![item_id])?;

    let remaining_ids = fetch_checklist_ids_for_task(conn, &task_id)?;
    let mut update_stmt = conn.prepare_cached(
        "UPDATE task_checklist_items
         SET position = ?1, version = ?2, updated_at = ?3
         WHERE id = ?4",
    )?;
    for (index, remaining_item_id) in remaining_ids.iter().enumerate() {
        let version = hlc.next_version_string();
        update_stmt.execute(params![index as i64, version, now, remaining_item_id])?;
    }

    let after = load_enriched_task_json(conn, &task_id)?;
    let mut item_sync_changes = Vec::with_capacity(remaining_ids.len() + 1);
    item_sync_changes.push(ChecklistItemSyncChange {
        item_id: item_id.to_string(),
        operation: ChecklistSyncOperation::Delete,
        snapshot: deleted_snapshot,
    });
    item_sync_changes.extend(remaining_ids.iter().map(item_upsert_change));

    Ok(TaskChecklistMutationResult {
        task_id: task_id.to_string(),
        before_task: before,
        after_task: after,
        summary: format!("Removed checklist item '{text}' for '{title}'"),
        item_sync_changes,
    })
}

pub fn reorder_task_checklist_items(
    conn: &Connection,
    hlc: &HlcSession<'_>,
    input: ReorderTaskChecklistItemsInput,
) -> Result<TaskChecklistMutationResult, StoreError> {
    let ReorderTaskChecklistItemsInput { task_id, item_ids } = input;
    let before = load_enriched_task_json(conn, &task_id)?;
    let title = task_title(&before).to_string();
    let existing_ids = fetch_checklist_ids_for_task(conn, &task_id)?;

    if item_ids.len() != existing_ids.len() {
        return Err(StoreError::Validation(format!(
            "reorder_task_checklist_items requires exactly {} ids for task '{task_id}', got {}",
            existing_ids.len(),
            item_ids.len()
        )));
    }

    let existing_set: HashSet<&str> = existing_ids.iter().map(ChecklistItemId::as_str).collect();
    let requested_set: HashSet<&str> = item_ids.iter().map(ChecklistItemId::as_str).collect();
    if existing_set != requested_set || requested_set.len() != item_ids.len() {
        return Err(StoreError::Validation(format!(
            "reorder_task_checklist_items must contain every checklist item for task '{task_id}' exactly once"
        )));
    }

    let now = lorvex_domain::sync_timestamp_now();
    let parent_version = hlc.next_version_string();
    touch_task_lww(conn, &task_id, &parent_version, &now)?;
    let mut update_stmt = conn.prepare_cached(
        "UPDATE task_checklist_items
         SET position = ?1, version = ?2, updated_at = ?3
         WHERE id = ?4 AND task_id = ?5",
    )?;
    for (index, item_id) in item_ids.iter().enumerate() {
        let version = hlc.next_version_string();
        update_stmt.execute(params![index as i64, version, now, item_id, task_id])?;
    }

    let after = load_enriched_task_json(conn, &task_id)?;
    let item_sync_changes = item_ids.iter().map(item_upsert_change).collect();

    Ok(TaskChecklistMutationResult {
        task_id: task_id.to_string(),
        before_task: before,
        after_task: after,
        summary: format!("Reordered checklist items for '{title}'"),
        item_sync_changes,
    })
}
